### DISCUSSION TITLE
Can't build the CS_Basic Console.

### DISCUSSION THREAD

**Discussion by spiroza** [mpaperno/WASimCommander#10] [Q&A]
I get...

Severity	Code	Description	Project	File	Line	Suppression State
Error	CS0246	The type or namespace name 'WASimCommander' could not be found (are you missing a using directive or an assembly reference?)	CS_BasicConsole	F:\MyMSFSProjects\WASimCommander-main\src\Testing\CS_BasicConsole\Program.cs	22	Active

Is there a preliminary step I'm missing?

Thanks!

**Comment by mpaperno** [comment]
Hi @spiroza ,

The error seems to indicate that the `WASimCommander.WASimClient.dll` is missing/can't be located.

Do you have the whole VS solution at that file path?  It's a bit strange because when that project is built, it should automatically build all the dependencies.  But maybe that's not working?

There should be a `src/build` folder after trying to build any of the projects... is there a `src\build\WASimClient_CLI` folder in there?  And inside one of the subfolders of that should be the dll file.

Which build type(s) are you trying (Release, Release-nnn, Debug)?

Thanks,
-Max

**Reply by spiroza** [reply]
I got it by building the whole project and then specifically building WASimClient_CLI.  Then I debugged Program.cs by right clicking on it in the explorer.  I'm stepping through it in the debugger now.  I have a lot to figure out and I'm thinking about how the same logic can be put in a C# Windows app.  I'm guessing that the LogHandler, ClientStatusHandler, ListResultsHandler and DataSubscriptionHandler would all be members of my MainWindow object.  Does one of those receive a "Sim Disconnected" event?

**Reply by mpaperno** [reply]
Cool, glad you got that figured. I'll check the dependencies again (it should have built the rest of the solution projects as needed, which would be WASimClient and WASimClient_CLI).

Yes, depending on what you want to do of course, but it would make sense to have those event handlers as part of one of your app's classes. They don't have to go in MainWindow per sé (eg. you could have a "Simulator" class which uses the WASimClient), but that's probably simplest to start with.

The `ClientStatusHandler()` would receive disconnection events, both when explicitly disconnected (calling `client.Disconnect[Server|Simulator]()`) and if the Sim itself quits.  The `ClientEvent` object it receives is described [here](https://mpaperno.github.io/WASimCommander/namespace_w_a_sim_commander_1_1_client.html#struct_w_a_sim_commander_1_1_client_1_1_client_event).

The C# interface could use some more [documentation](https://mpaperno.github.io/WASimCommander/class_w_a_sim_commander_1_1_c_l_i_1_1_client_1_1_w_a_sim_client.html) specific to it.  In general it is all basically identical to the C++ version, except of course some of the variable types (`String` vs `std::string` for example), and the event handlers (vs. callbacks in C++).

Holler with any questions!

Cheers,
-Max

**Comment by spiroza** [comment]
You may be right about that extra step.  I haven't used Visual Studio in a
few years and definitely haven't had any mixed language solutions so there
were some things I had to figure out...e.g. right clicking on the specific
file to debug it so I could step through it.  It will be a challenge for me
to get this all connected to a window.  If I run into trouble I'll be sure
to beg for help.

Thank you!

**Reply by mpaperno** [reply]
Not sure if this is what you ran into, but to simplify running/debugging a project in VS you can r-click on it in the solution explorer and select "Set as Startup Project."

Also I'd create a new solution with just your CS project. You just need the `WASimCommander.WASimClient.dll` to link to... copy it to a "lib" folder in your project, along with `Ljwhost.dll` and `client_conf.ini`.  Then add something like this to your CS project file to reference the DLL and copy it to the final build directory:

```xml

  <!-- WASimCommander Client -->
  <ItemGroup >
    <Reference Include="WASimCommander.WASimClient, Culture=neutral, processorArchitecture=MSIL" Version="1.1.0">
      <HintPath>.\lib\WASimCommander\WASimCommander.WASimClient.dll</HintPath>
      <Private>true</Private>
      <EmbedInteropTypes>false</EmbedInteropTypes>
    </Reference>
  </ItemGroup>
  <ItemGroup>
    <ContentWithTargetPath Include=".\lib\WASimCommander\Ijwhost.dll">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <TargetPath>Ijwhost.dll</TargetPath>
    </ContentWithTargetPath>
    <ContentWithTargetPath Include=".\lib\WASimCommander\client_conf.ini">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <TargetPath>client_conf.ini</TargetPath>
    </ContentWithTargetPath>
  </ItemGroup>
```

Also, this is not a UI project but maybe more examples will help... I'm using the Client in this project: [MSFSTouchPortalPlugin](https://github.com/mpaperno/MSFSTouchPortalPlugin/tree/next/MSFSTouchPortalPlugin) and more specifically in the `Services/SimConnectService.cs` class.  (This project originally used SimConnect only, the WASimCommander features are added as an option for any users who install the corresponding WASM module.)

HTH,
-Max
